### ISSUE TITLE
MiniLLM: `_compute_advantage` discounts by absolute position, scaling every advantage by gamma^t

### ISSUE THREAD

**Issue by MehrdadMoghimi** [huggingface/trl#6626]
### Reproduction

`MiniLLMTrainer._compute_advantage` builds `gamma_pow[i] = gamma**i` from the **absolute** index `i`, then does a reverse cumsum. That computes `sum_{i>=t} gamma^i * R_i = gamma^t * A_t` instead of the documented `sum_{i>=t} gamma^(i-t) * R_i = A_t`, so every position is scaled by an extra `gamma^t`.

**No exception is raised** — with `length_normalization=False` the advantages are silently wrong (later tokens are exponentially suppressed toward zero). With `length_normalization=True` the `gamma^t` cancels between numerator and denominator, but `gamma**i` underflows to exactly `0.0` in float32, giving `0/0 = nan` on long completions.

The repro calls the method with a stub `self` because it reads only `self.gamma` and `self.length_normalization`.

```python
import types

import torch

from trl.experimental.minillm.minillm_trainer import MiniLLMTrainer

advantage = MiniLLMTrainer._compute_advantage  # real method, called with a stub `self`

# Constant reward of 1.0 at every position: rewards = teacher - student = 1 - 0 = 1
T, GAMMA = 512, 0.9
student, teacher, mask = torch.zeros(1, T), torch.ones(1, T), torch.ones(1, T)

print("BUG 1 (length_normalization=False): every position is scaled by an extra gamma^t.")
print(f"With a constant reward of 1.0, the advantage should be A_t = (1-{GAMMA}^(T-t))/(1-{GAMMA}) = 10.0")
print("at every position. Instead it decays, and the error is exactly a factor of gamma^t:\n")
out = advantage(types.SimpleNamespace(gamma=GAMMA, length_normalization=False), student, teacher, mask)[0]
print(f"{'t':>5} {'expected':>12} {'actual':>14} {'actual/expected':>17} {'gamma^t':>14}")
for t in [0, 50, 100, 200, 400]:
    expected = (1 - GAMMA ** (T - t)) / (1 - GAMMA)
    print(f"{t:>5} {expected:>12.4f} {out[t].item():>14.6e} {out[t].item() / expected:>17.6e} {GAMMA**t:>14.6e}")
print("\n^ the last two columns match: gamma_pow is indexed by the absolute i, not by (i - t).\n")

print("BUG 2 (length_normalization=True): the gamma^t factor cancels between numerator and")
print("denominator, so every position should be exactly 1.0 -- until gamma^i underflows to 0.0")
print("in float32. Past that point both sums are 0.0 and the division yields 0/0 = nan:\n")
for gamma, T2 in [(0.5, 512), (0.9, 1024), (0.95, 2048), (0.99, 16384)]:
    out = advantage(types.SimpleNamespace(gamma=gamma, length_normalization=True), torch.zeros(1, T2),
                    torch.ones(1, T2), torch.ones(1, T2))[0]
    nan_at = (~torch.isfinite(out)).nonzero()
    zero_at = (torch.pow(gamma, torch.arange(T2)) == 0.0).nonzero()
    print(f"  gamma={gamma:<5} T={T2:<5} A_0={out[0].item():<5} A_last={out[-1].item():<5} "
          f"first nan at t={str(int(nan_at[0, 0])) if len(nan_at) else 'never':<7} "
          f"first gamma^i == 0.0 at i={str(int(zero_at[0, 0])) if len(zero_at) else 'never'}")
print("\n^ the nan index always equals the underflow index. Smaller gamma underflows sooner.")
```

outputs:

```
BUG 1 (length_normalization=False): every position is scaled by an extra gamma^t.
With a constant reward of 1.0, the advantage should be A_t = (1-0.9^(T-t))/(1-0.9) = 10.0
at every position. Instead it decays, and the error is exactly a factor of gamma^t:

    t     expected         actual   actual/expected        gamma^t
    0      10.0000   9.999997e+00      9.999997e-01   1.000000e+00
   50      10.0000   5.153767e-02      5.153767e-03   5.153775e-03
  100      10.0000   2.656132e-04      2.656132e-05   2.656140e-05
  200      10.0000   7.055040e-09      7.055040e-10   7.055079e-10
  400       9.9999   4.977323e-18      4.977360e-19   4.977414e-19

^ the last two columns match: gamma_pow is indexed by the absolute i, not by (i - t).

BUG 2 (length_normalization=True): the gamma^t factor cancels between numerator and
denominator, so every position should be exactly 1.0 -- until gamma^i underflows to 0.0
in float32. Past that point both sums are 0.0 and the division yields 0/0 = nan:

  gamma=0.5   T=512   A_0=1.0   A_last=nan   first nan at t=150     first gamma^i == 0.0 at i=150
  gamma=0.9   T=1024  A_0=1.0   A_last=nan   first nan at t=987     first gamma^i == 0.0 at i=987
  gamma=0.95  T=2048  A_0=1.0   A_last=nan   first nan at t=2028    first gamma^i == 0.0 at i=2028
  gamma=0.99  T=16384 A_0=1.0   A_last=nan   first nan at t=10346   first gamma^i == 0.0 at i=10346

^ the nan index always equals the underflow index. Smaller gamma underflows sooner.
```

The referenced implementation ([LMOps `losses.py#L37-L49`](https://github.com/microsoft/LMOps/blob/e210d2c026b9958617887762400778ace81172e6/minillm/minillm/losses.py#L37-L49)) uses a backward recurrence, `last_rw = rewards[:, t] + gamma * last_rw`, which produces the relative exponent by construction and never materializes `gamma**i` — so it has neither problem. Porting that recurrence would fix both at once.

Happy to submit a PR porting the backward recurrence.

### System Info

```
- Platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
- Python version: 3.12.4
- TRL version: 1.10.0.dev0+aa59202
- PyTorch version: 2.13.0
- accelerator(s): NVIDIA GeForce RTX 4090
- Transformers version: 5.14.1
- Accelerate version: 1.14.0
- Accelerate config: not found
- Datasets version: 5.0.1
- HF Hub version: 1.26.0
- bitsandbytes version: 0.50.0
- DeepSpeed version: 0.19.3
- Liger-Kernel version: 0.8.0
- PEFT version: 0.20.0
- vLLM version: not installed
```

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by rp440** [comment]
I reproduced this on CPU on Apple Silicon using the real MiniLLMTrainer._compute_advantage method. With four unit rewards, gamma=0.5, and length_normalization=False, TRL returns [1.875, 0.875, 0.375, 0.125] instead of [1.875, 1.75, 1.5, 1.0]. With gamma=0.9, length_normalization=True, and a sequence length of 1024, the first non-finite advantage appears at position 987.

A backward recurrence fixes both the absolute-position scaling and the long-sequence underflow:
`running_advantage = rewards[:, t] + gamma * running_advantage`

One scope detail I noticed is that the original Microsoft MiniLLM implementation divides by the ordinary number of remaining valid tokens, while TRL’s _compute_advantage docstring specifies a discounted normalization denominator. To avoid an unrelated behavior change, I would preserve TRL’s documented semantics and apply the backward recurrence to both the advantage numerator and the discounted normalization denominator.

Since you mentioned being happy to submit a PR, are you already working on it? If not, I would be happy to implement the fix
